driver-core/src/test/functional/com/mongodb/client/model/expressions/TypeExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,247 @@
+    }
+
+    @Test
+    public void isStringOrTest() {
+        assertExpression(
+                "abc",
+                of("abc").isStringOr(of("or")),
+                "{'$cond': [{'$eq': [{'$type': 'abc'}, 'string']}, 'abc', 'or']}");
+        // non-string:
+        assertExpression("or", ofIntegerArray(1).isStringOr(of("or")));
+        assertExpression("or", ofNull().isStringOr(of("or")));
+    }
+
+    @Test
+    public void isDateOrTest() {
+        Instant date = Instant.parse("2007-12-03T10:15:30.005Z");
+        assertExpression(
+                Instant.ofEpochMilli(1196676930005L),
+                of(date).isDateOr(of(date.plusMillis(10))),
+                "{'$cond': [{'$in': [{'$type': {'$date': '2007-12-03T10:15:30.005Z'}}, "
+                        + "['date', 'timestamp']]}, {'$date': '2007-12-03T10:15:30.005Z'}, "
+                        + "{'$date': '2007-12-03T10:15:30.015Z'}]}");
+        // non-date:
+        assertExpression(date, ofIntegerArray(1).isDateOr(of(date)));
+        assertExpression(date, ofNull().isDateOr(of(date)));
+    }
+
+    @Test
+    public void isArrayOrTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/isArray/ (36 |47)
+        assertExpression(
+                Arrays.asList(1, 2),
+                ofIntegerArray(1, 2).isArrayOr(ofIntegerArray(99)),
+                "{'$cond': [{'$isArray': [[1, 2]]}, [1, 2], [99]]}");
+        // non-array:
+        assertExpression(Arrays.asList(1, 2), of(true).isArrayOr(ofIntegerArray(1, 2)));
+        assertExpression(Arrays.asList(1, 2), ofNull().isArrayOr(ofIntegerArray(1, 2)));
+    }
+
+    @Test
+    public void isDocumentOrTest() {
+        BsonDocument doc = BsonDocument.parse("{a: 1}");
+        assertExpression(doc,
+                of(doc).isDocumentOr(of(BsonDocument.parse("{b: 2}"))),
+                "{'$cond': [{'$eq': [{'$type': {'$literal': {'a': 1}}}, 'object']}, "
+                        + "{'$literal': {'a': 1}}, {'$literal': {'b': 2}}]}");
+        // non-document:
+        assertExpression(doc, ofIntegerArray(1).isDocumentOr(of(doc)));
+        assertExpression(doc, ofNull().isDocumentOr(of(doc)));
+    }
+
+    // conversions
+    // https://www.mongodb.com/docs/manual/reference/operator/aggregation/convert/ (38 |40)
+    // Convert is not implemented: too dynamic, conversions should be explicit.
+
+    /*
+    Useful conversions:
+    - anything-string: toString to a parsable type
+    - string-parse-anything: every type should allow parsing to a string
+      - presently excludes objects (docs) and arrays
+      - includes formatted date strings
+    - milliseconds since epoch to date
+
+    Convert also defines many conversions that do not seem useful:
+    - boolean-number conversions: t/f to-from 1/0
+    - boolean-other (oid/str/date) - always true, broken for strings; pointless
+    - number-number - "underlying" json type is changed, possible exceptions?
+     */
```

Comment:
[optional]
This comment looks like some thoughts that may be useful to you, but I am not sure this is useful to others, especially future readers (I can't really understand what the comment says and why it is here). The same is applicable to the comment in the beginning of this test class.

I am rising this concern in case others have similar thoughts about such comments. If that's the case, maybe we should consider removing the comments / writing them in a way that is more clear to readers  / putting them on the relevant program elements rather than having them hanging in the air.

Reply:
I will try to make them more clear. They are usually explanations for why something is one way or another, or where to find what you might be looking for, or why you can't find something you expected to. I don't think they belong on our external API, and MqlEx does not seem like a good fit either for this case.